SegFault on responding to ipfs cat

#### Version information:
go-ipfs version: 0.5.0-dev-4acb15d52-dirty
Repo version: 7
System version: arm/linux
Golang version: go1.14

#### Description:
Setup is described in #301 
I was testing the latest version of Bitswap, 0.2.3, in the latest state of IPFS. I think this is the right place for this bug, but please redirect me if not.

Doing step 5. on **node B** caused a segfault error on **node A**, and consequently a daemon crash on **node A**.

The previous IPFS binary I was running on the nodes was from February, Bitswap 0.1.11, and everything was working.

<details>
  <summary>Previously tested working version</summary>
<p>

```
go-ipfs version: 0.5.0-dev-a22dc826c-dirty
Repo version: 7
System version: arm/linux
Golang version: go1.13.8
```

</p>
</details>

I also submit here the relevant logs. Apologies regarding the format but couldn't find an easy way to make it softer on the eyes.
<details>
  <summary>Log from Node A</summary>
<p>

```
1970-01-12T04:24:02.697Z	DEBUG	bitswap_network	network/ipfs_impl.go:268	bitswap net handleNewStream from Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E
1970-01-12T04:24:02.697Z	DEBUG	engine	decision/engine.go:515	Bitswap engine <- msg	{"local": "QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6", "from": "Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E", "entryCount": 1}
1970-01-12T04:24:02.698Z	DEBUG	engine	decision/engine.go:519	Bitswap engine <- want-have	{"local": "QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6", "from": "Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E", "cid": "QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk"}
1970-01-12T04:24:02.698Z	DEBUG	engine	decision/ledger.go:95	peer Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E wants QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk
1970-01-12T04:24:02.698Z	DEBUG	engine	decision/engine.go:615	Bitswap engine: block found	{"local": "QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6", "from": "Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E", "cid": "QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk", "isWantBlock": true}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x11c70]

goroutine 16009 [running]:
runtime/internal/atomic.goXadd64(0x34935ac, 0x1, 0x0, 0x38b2540, 0x22)
	/usr/local/go/src/runtime/internal/atomic/atomic_arm.go:103 +0x1c
github.com/ipfs/go-bitswap/network.(*impl).handleNewStream(0x3493560, 0x10c1038, 0x5ed24e0)
	pkg/mod/github.com/ipfs/go-bitswap@v0.2.3/network/ipfs_impl.go:270 +0x2b4
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).SetStreamHandler.func1(0x3685a60, 0x13, 0x66cc8440, 0x5ed24e0, 0x1, 0x0)
	pkg/mod/github.com/libp2p/go-libp2p@v0.6.0/p2p/host/basic/basic_host.go:397 +0x7c
created by github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler
	pkg/mod/github.com/libp2p/go-libp2p@v0.6.0/p2p/host/basic/basic_host.go:295 +0x55c
```

</p>
</details>

<details>
  <summary>Log from Node B</summary>
<p>

```
1970-01-12T04:23:58.896Z	DEBUG	cmds/http	http/handler.go:88	incoming API request: /cat?arg=QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk&encoding=json&stream-channels=true
1970-01-12T04:23:58.901Z	DEBUG	blockservice	go-blockservice@v0.1.2/blockservice.go:238	Blockservice: Searching bitswap
1970-01-12T04:23:58.901Z	DEBUG	bitswap	getter/getter.go:80	Bitswap.GetBlockRequest.Start	{"cid": "QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk"}
1970-01-12T04:23:58.902Z	INFO	bs:sess	session/session.go:439	Ses1: No peers - broadcasting 1 want HAVE requests

1970-01-12T04:23:58.902Z	DEBUG	bitswap	wantmanager/wantmanager.go:81	BroadcastWantHaves session1: [QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk]
1970-01-12T04:23:58.903Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/swarm.go:292	[Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E] opening stream to peer [QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6]
1970-01-12T04:23:58.904Z	DEBUG	bitswap	messagequeue/messagequeue.go:473	Bitswap -> want-have	{"local": "Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E", "to": "QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6", "cid": "QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk"}
1970-01-12T04:23:59.902Z	DEBUG	bitswap	wantmanager/wantmanager.go:81	BroadcastWantHaves session1: [QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk]
1970-01-12T04:23:59.902Z	DEBUG	bs:sess	session/session.go:344	Ses1: FindMorePeers with want QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk (1st of 1 wants)
1970-01-12T04:23:59.902Z	DEBUG	bitswap	providerquerymanager/providerquerymanager.go:323	New Provider Query on cid: QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk
1970-01-12T04:23:59.903Z	DEBUG	bitswap	providerquerymanager/providerquerymanager.go:230	Beginning Find Provider Request for cid: QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk
1970-01-12T04:23:59.915Z	DEBUG	basichost	basic/basic_host.go:587	host Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E dialing QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6
1970-01-12T04:23:59.915Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:193	[Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E] swarm dialing peer [QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6]
1970-01-12T04:23:59.917Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:121	InterfaceAddresses: from manet:[/ip4/127.0.0.1 /ip4/192.168.21.42 /ip4/10.0.21.42 /ip6/::1 /ip6/fd69:791a:f930::1 /ip6/fe80::2d0:12ff:fe55:f1d3 /ip6/fe80::260a:64ff:fe1c:866e]
1970-01-12T04:23:59.918Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:133	InterfaceAddresses: usable:[/ip4/127.0.0.1 /ip4/192.168.21.42 /ip4/10.0.21.42 /ip6/::1 /ip6/fd69:791a:f930::1]
1970-01-12T04:23:59.918Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:64	adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/127.0.0.1/tcp/4001 [/ip4/127.0.0.1/tcp/4001]
1970-01-12T04:23:59.918Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:64	adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/192.168.21.42/tcp/4001 [/ip4/127.0.0.1/tcp/4001 /ip4/192.168.21.42/tcp/4001]
1970-01-12T04:23:59.918Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:64	adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/10.0.21.42/tcp/4001 [/ip4/127.0.0.1/tcp/4001 /ip4/192.168.21.42/tcp/4001 /ip4/10.0.21.42/tcp/4001]
1970-01-12T04:23:59.919Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:64	adding resolved addr:/ip6/::/tcp/4001 /ip6/::1/tcp/4001 [/ip6/::1/tcp/4001]
1970-01-12T04:23:59.919Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:64	adding resolved addr:/ip6/::/tcp/4001 /ip6/fd69:791a:f930::1/tcp/4001 [/ip6/::1/tcp/4001 /ip6/fd69:791a:f930::1/tcp/4001]
1970-01-12T04:23:59.919Z	DEBUG	addrutil	go-addr-util@v0.0.1/addr.go:109	ResolveUnspecifiedAddresses:[/p2p-circuit /ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] [/ip4/127.0.0.1 /ip4/192.168.21.42 /ip4/10.0.21.42 /ip6/::1 /ip6/fd69:791a:f930::1] [/p2p-circuit /ip4/127.0.0.1/tcp/4001 /ip4/192.168.21.42/tcp/4001 /ip4/10.0.21.42/tcp/4001 /ip6/::1/tcp/4001 /ip6/fd69:791a:f930::1/tcp/4001]
1970-01-12T04:23:59.919Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:371	Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E swarm dialing QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6
1970-01-12T04:23:59.919Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:192	[limiter] adding a dial job through limiter: /ip4/10.0.21.35/tcp/4001
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:160	[limiter] taking FD token: peer: QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip4/10.0.21.35/tcp/4001; prev consuming: 0
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:166	[limiter] executing dial; peer: QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip4/10.0.21.35/tcp/4001; FD consuming: 1; waiting: 0
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:192	[limiter] adding a dial job through limiter: /ip6/fda8:d807:3852::1/tcp/4001
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:160	[limiter] taking FD token: peer: QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip6/fda8:d807:3852::1/tcp/4001; prev consuming: 1
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:166	[limiter] executing dial; peer: QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip6/fda8:d807:3852::1/tcp/4001; FD consuming: 2; waiting: 0
1970-01-12T04:23:59.920Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:455	Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E swarm dialing QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6 /ip6/fda8:d807:3852::1/tcp/4001
1970-01-12T04:23:59.921Z	DEBUG	reuseport-transport	go-reuseport-transport@v0.0.2/reuseport.go:60	failed to reuse port, dialing with a random port: dial tcp6 [::]:4001->[fda8:d807:3852::1]:4001: connect: permission denied
1970-01-12T04:23:59.921Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:82	[limiter] freeing FD token; waiting: 0; consuming: 2
1970-01-12T04:23:59.921Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:109	[limiter] freeing peer token; peer QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip6/fda8:d807:3852::1/tcp/4001; active for peer: 2; waiting on peer limit: 0
1970-01-12T04:23:59.922Z	INFO	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:420	got error on dial: dial tcp6 [fda8:d807:3852::1]:4001: connect: permission denied
1970-01-12T04:23:59.922Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:455	Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E swarm dialing QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6 /ip4/10.0.21.35/tcp/4001
1970-01-12T04:24:00.903Z	DEBUG	bitswap	wantmanager/wantmanager.go:81	BroadcastWantHaves session1: [QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk]
1970-01-12T04:24:02.903Z	DEBUG	bitswap	wantmanager/wantmanager.go:81	BroadcastWantHaves session1: [QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk]
1970-01-12T04:24:03.886Z	DEBUG	mdns	discovery/mdns.go:133	starting mdns query
1970-01-12T04:24:03.888Z	DEBUG	mdns	discovery/mdns.go:159	Handling MDNS entry: 10.0.21.42:4001 Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E
1970-01-12T04:24:03.889Z	DEBUG	mdns	discovery/mdns.go:167	got our own mdns entry, skipping
1970-01-12T04:24:03.938Z	DEBUG	bootstrap	bootstrap/bootstrap.go:142	Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E no more bootstrap peers to create 4 connections
1970-01-12T04:24:03.938Z	DEBUG	bootstrap	bootstrap/bootstrap.go:90	Qmai3JBnAJ3Ahzizgod69bwrpcMEys4tfntAV4tbjGWz2E bootstrap error: not enough bootstrap peers to bootstrap
1970-01-12T04:24:03.938Z	WARN	p2pnode	libp2p/pnet.go:56	We are in private network and have no peers.
1970-01-12T04:24:03.938Z	WARN	p2pnode	libp2p/pnet.go:57	This might be configuration mistake.
1970-01-12T04:24:04.922Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:82	[limiter] freeing FD token; waiting: 0; consuming: 1
1970-01-12T04:24:04.922Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:109	[limiter] freeing peer token; peer QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6; addr: /ip4/10.0.21.35/tcp/4001; active for peer: 1; waiting on peer limit: 0
1970-01-12T04:24:04.922Z	INFO	swarm2	go-libp2p-swarm@v0.2.2/swarm_dial.go:420	got error on dial: dial tcp4 0.0.0.0:4001->10.0.21.35:4001: i/o timeout
1970-01-12T04:24:04.923Z	DEBUG	swarm2	go-libp2p-swarm@v0.2.2/limiter.go:200	[limiter] clearing all peer dials: QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6
1970-01-12T04:24:04.923Z	DEBUG	bitswap	providerquerymanager/providerquerymanager.go:242	failed to connect to provider QmXa8Lgbi8VoeZeAGEumEEdPhMXfstGjDoBwMsTogXrSu6: failed to dial : all dials failed
  * [/ip6/fda8:d807:3852::1/tcp/4001] dial tcp6 [fda8:d807:3852::1]:4001: connect: permission denied
  * [/ip4/10.0.21.35/tcp/4001] dial tcp4 0.0.0.0:4001->10.0.21.35:4001: i/o timeout
1970-01-12T04:24:04.923Z	DEBUG	bitswap	providerquerymanager/providerquerymanager.go:323	Finished Provider Query on cid: QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk
1970-01-12T04:24:05.903Z	DEBUG	bitswap	wantmanager/wantmanager.go:81	BroadcastWantHaves session1: [QmRKURFp8u41UANSi8pK1UDEKx2FHSdsvVHxXwxwehDnEk]
1970-01-12T04:24:08.544Z	DEBUG	cmds	go-ipfs-cmds@v0.1.2/command.go:78	error occured in call, closing with error: context canceled
```
</p>
</details>

Don't focus too much on the timestamps, they aren't synced between nodes.

Do the devs have any idea of what caused this, or should I look into it? Thank you.

Edit: Fixed log formatting.
Edit: Added previously working version's details

--- a/network/ipfs_impl.go
+++ b/network/ipfs_impl.go
@@ -67,6 +67,10 @@ func processSettings(opts ...NetOpt) Settings {
 // impl transforms the ipfs network interface, which sends and receives
 // NetMessage objects, into the bitswap network interface.
 type impl struct {
+	// NOTE: Stats must be at the top of the heap allocation to ensure 64bit
+	// alignment.
+	stats Stats
+
 	host    host.Host
 	routing routing.ContentRouting
 
@@ -79,8 +83,6 @@ type impl struct {
 
 	// inbound messages from the network are forwarded to the receiver
 	receiver Receiver
-
-	stats Stats
 }
 
 type streamMessageSender struct {

